Sources/NIOSOCKS/Channel Handlers/SOCKSClientAuthenticationDelegate.swift

```diff
@@ -0,0 +1,67 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+public struct UnexpectedAuthenticationMethod: Error {
+    public var expected: [AuthenticationMethod]
+    public var actual: AuthenticationMethod
+    
+    public init(expected: [AuthenticationMethod], actual: AuthenticationMethod) {
+        self.expected = expected
+        self.actual = actual
+    }
+}
+
+public enum AuthenticationResult: Hashable {
+    case needsMoreData
```

Comment:
I was trying to make this future proof. For example if `channelRead` only delivered half of the data it's a way of waiting for another read to deliver the rest. This could become useful for more complicated (particularly custom) authentication methods.

Reply:
I think this is not the kind of future proofing we want to build: it forces us to push all the complexity into adopter code, rather than handling it ourselves. I think the implementation here should know how the auth methods work, and should only ask the user for things it cannot get itself, e.g. username/password.

Reply:
There are custom/private authentication methods we need to somehow support though. Take a look at page 3 https://datatracker.ietf.org/doc/html/rfc1928

Reply:
I would be a _substantial_ sum of money that supporting them is never needed.

Reply:
Additionally, we can always add that support down the road if we have to by supporting multiple kinds of delegates.

Reply:
Where are we with this?

Reply:
I agree that the completely custom methods we can probably forget, but I'm a little worried about not providing support for particularly IANA registered methods. Won't it break the public API if we move to do this later?

Reply:
No, we can add new delegate types that can do _more_ later so long as they don't do less.

Tbqh given that we don't support auth _at all_ at the moment, the best design may be to define away almost all of the auth stuff: support only "no auth", don't have any entry points that take delegates, just remove all of it.

We can safely add additional API entry points when we come to wanting to support auth.

Reply:
I meant adding this `.needsMoreData` enum case to `AuthenticationResult`. I'm fine with removing auth until someone requests it. I'll purge.

Reply:
All sorted, I'll resolve the convo